### PR TITLE
Allow other auth providers (GCP, OIDC, Azure, OpenStack)

### PR DESCRIPTION
**Pull request by huguesalary**
This PR allows the descheduler to be used with other auth providers (Issue #93), however, the branch in its current state does not compile.

Running `glide update` pulls a dependency that is breaking the build.

Removing `vendor/k8s.io/kubernetes/vendor/` seems to fix the issue.

```
# github.com/kubernetes-incubator/descheduler/pkg/utils
pkg/utils/predicates.go:84:71: cannot use req.MatchExpressions (type []"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".NodeSelectorRequirement) as type []"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/vendor/k8s.io/api/core/v1".NodeSelectorRequirement in argument to helper.NodeSelectorRequirementsAsSelector
# github.com/kubernetes-incubator/descheduler/pkg/descheduler/pod
pkg/descheduler/pod/pods.go:100:28: cannot use pod (type *"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Pod) as type *"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/vendor/k8s.io/api/core/v1".Pod in argument to "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/pkg/kubelet/types".IsCriticalPod
pkg/descheduler/pod/pods.go:104:22: cannot use pod (type *"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Pod) as type *"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/vendor/k8s.io/api/core/v1".Pod in argument to qos.GetPodQOS
pkg/descheduler/pod/pods.go:104:28: invalid operation: qos.GetPodQOS(pod) == "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".PodQOSBestEffort (mismatched types "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/vendor/k8s.io/api/core/v1".PodQOSClass and "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".PodQOSClass)
pkg/descheduler/pod/pods.go:108:22: cannot use pod (type *"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Pod) as type *"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/vendor/k8s.io/api/core/v1".Pod in argument to qos.GetPodQOS
pkg/descheduler/pod/pods.go:108:28: invalid operation: qos.GetPodQOS(pod) == "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".PodQOSBurstable (mismatched types "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/vendor/k8s.io/api/core/v1".PodQOSClass and "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".PodQOSClass)
pkg/descheduler/pod/pods.go:112:22: cannot use pod (type *"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".Pod) as type *"github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/vendor/k8s.io/api/core/v1".Pod in argument to qos.GetPodQOS
pkg/descheduler/pod/pods.go:112:28: invalid operation: qos.GetPodQOS(pod) == "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".PodQOSGuaranteed (mismatched types "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/kubernetes/vendor/k8s.io/api/core/v1".PodQOSClass and "github.com/kubernetes-incubator/descheduler/vendor/k8s.io/api/core/v1".PodQOSClass)
make: *** [build] Error 2
```

### CODE REVIEW

**Comment by aveshagarwal** [comment]
Can one of the admins verify this patch?

**Comment by ravisantoshgudimetla** [comment]
@huguesalary Thanks for the PR, can you run `glide update -v` which removes nested vendor dependencies?

**Comment by huguesalary** [comment]
Done.

Now everything's compiling correctly. This might be something to add to the README.

Cheers

**Comment by ravisantoshgudimetla** [comment]
Sorry for going back and forth. Can you split the commits into 2. One with original change and other with generated file.
/cc @aveshagarwal @ingvagabund

**Comment by huguesalary** [comment]
Do you mean:

1 commit with `glide.yaml` and `pkg/descheduler/client/client.go`
1 commit with the rest (`glide.lock` and `vendor/`) ?

**Comment by ravisantoshgudimetla** [comment]
> Do you mean:

> 1 commit with glide.yaml and pkg/descheduler/client/client.go
> 1 commit with the rest (glide.lock and vendor/) ?

That is correct, it would be easy to review as well.

**Comment by huguesalary** [comment]
No problem, here it is

**Comment by ravisantoshgudimetla** [comment]
Has this been tested against GKE @huguesalary. If yes, this LGTM as it has passed e2e's as well.

**Comment by aveshagarwal** [comment]
@huguesalary thanks for this PR. Since descheduler still points to kube 1.9, I'd recommend to "k8s.io/client-go/plugin/pkg/client/auth" and any other related dependencies to keep at 1.9 too to avoid updating to latest kube. Any dependency outside of kube could be updated to latest or whatever i think if really needed.

**Comment by huguesalary** [comment]
> Has this been tested against GKE @huguesalary. If yes, this LGTM as it has passed e2e's as well.

Yup! 

>  Since descheduler still points to kube 1.9, I'd recommend to "k8s.io/client-go/plugin/pkg/client/auth" and any other related dependencies to keep at 1.9 too to avoid updating to latest kube. Any dependency outside of kube could be updated to latest or whatever i think if really needed.

Sorry, I'm not exactly sure I understand. 

Happy to modify the PR though.

**Comment by aveshagarwal** [comment]
> Sorry, I'm not exactly sure I understand.

Nevermind. I checked now and your PR is not updating anything in kube. Will review it in detail tomorrow. Again thanks for this PR.

**Comment by ingvagabund** [comment]
I have always thought the `glide.yaml` and `glide.lock` files are in sync, i.e. they contain the same list of dependencies. Just the `glide.lock` is more machine oriented so it contains more detailed information. Currently, I see the `glide.yml` contains less dependencies than the `glide.lock`.

File: glide.yaml [hunk 1]
```diff
@@ -18,6 +18,8 @@ import:
 - package: k8s.io/gengo
 - package: github.com/ugorji/go
   version: v.1.1-beta
+- package: github.com/Azure/go-autorest
+  version: e14a70c556c8e0db173358d1a903dca345a8e75e
 - package: golang.org/x/tools
   subpackages:
   - imports
```
Comment (ingvagabund) [inline glide.yaml]: Those are all the direct dependencies of the descheduler:

```sh
$ golist --package-path github.com/kubernetes-incubator/descheduler --imported
github.com/golang/glog
github.com/kubernetes-incubator/descheduler/cmd/descheduler/app
github.com/kubernetes-incubator/descheduler/cmd/descheduler/app/options
github.com/kubernetes-incubator/descheduler/pkg/api
github.com/kubernetes-incubator/descheduler/pkg/api/install
github.com/kubernetes-incubator/descheduler/pkg/api/v1alpha1
github.com/kubernetes-incubator/descheduler/pkg/apis/componentconfig
github.com/kubernetes-incubator/descheduler/pkg/apis/componentconfig/install
github.com/kubernetes-incubator/descheduler/pkg/apis/componentconfig/v1alpha1
github.com/kubernetes-incubator/descheduler/pkg/descheduler
github.com/kubernetes-incubator/descheduler/pkg/descheduler/client
github.com/kubernetes-incubator/descheduler/pkg/descheduler/evictions
github.com/kubernetes-incubator/descheduler/pkg/descheduler/evictions/utils
github.com/kubernetes-incubator/descheduler/pkg/descheduler/node
github.com/kubernetes-incubator/descheduler/pkg/descheduler/pod
github.com/kubernetes-incubator/descheduler/pkg/descheduler/scheme
github.com/kubernetes-incubator/descheduler/pkg/descheduler/strategies
github.com/kubernetes-incubator/descheduler/pkg/utils
github.com/spf13/cobra
github.com/spf13/pflag
github.com/ugorji/go/codec
k8s.io/api/core/v1
k8s.io/api/policy/v1beta1
k8s.io/apimachinery/pkg/api/errors
k8s.io/apimachinery/pkg/api/resource
k8s.io/apimachinery/pkg/apimachinery/announced
k8s.io/apimachinery/pkg/apimachinery/registered
k8s.io/apimachinery/pkg/apis/meta/v1
k8s.io/apimachinery/pkg/conversion
k8s.io/apimachinery/pkg/fields
k8s.io/apimachinery/pkg/labels
k8s.io/apimachinery/pkg/runtime
k8s.io/apimachinery/pkg/runtime/schema
k8s.io/apimachinery/pkg/runtime/serializer
k8s.io/apiserver/pkg/util/flag
k8s.io/apiserver/pkg/util/logs
k8s.io/client-go/kubernetes
k8s.io/client-go/listers/core/v1
k8s.io/client-go/plugin/pkg/client/auth
k8s.io/client-go/rest
k8s.io/client-go/tools/cache
k8s.io/client-go/tools/clientcmd
k8s.io/kubernetes/pkg/api/v1/resource
k8s.io/kubernetes/pkg/apis/core
k8s.io/kubernetes/pkg/apis/core/v1/helper
k8s.io/kubernetes/pkg/apis/core/v1/helper/qos
k8s.io/kubernetes/pkg/kubelet/types
k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/priorities/util
```

Would be worth minimizing the content of the vendor cause only `k8s.io/kubernetes/pkg/kubelet/types` package is used although the vendor container entire tree of `k8s.io/kubernetes/pkg/kubelet`.

Though, not a blocker for the PR.

**Comment by ingvagabund** [comment]
@aveshagarwal  @ravisantoshgudimetla I am fine with the PR as it is.

**Comment by aveshagarwal** [comment]
> I have always thought the glide.yaml and glide.lock files are in sync, i.e. they contain the same list of dependencies. Just the glide.lock is more machine oriented so it contains more detailed information. Currently, I see the glide.yml containing less dependencies than the glide.lock.

> EDIT: The glide project [1] does not say the glide.yml has to contain all the dependencies.

Right, basically glide.yaml only contains those dependencies that have been pinned to specific version/tag etc . Whereas glide.lock contains all dependencies and their versions.

**Comment by ingvagabund** [comment]
@ravisantoshgudimetla none here, go ahead :)

**Comment by aveshagarwal** [comment]
@ravisantoshgudimetla my only concern is that we should make sure that vendor update is actually what is really needed and there is nothing extra or manually added. As far as it is fine, lgtm to me too.

**Comment by huguesalary** [comment]
As far as I'm concerned, all the dependencies added by this PR were generated by adding `github.com/Azure/go-autorest` to `glide.yaml` then running `glide update -v`, so I would think they're all required.

**Comment by aveshagarwal** [comment]
@huguesalary thanks. Then I am fine with it.

/lgtm

**Comment by aveshagarwal** [comment]
All tests passed. Merging now.

**Pull request merged**